traveler-ios-ui/TravelerKitUI/ViewControllers/CancellationViewController.swift

```diff
@@ -15,50 +15,127 @@ protocol CancellationViewControllerDelegate: class {
+                cell.secondValueLabel?.text = quote?.totalRefund.localizedDescriptionInBaseCurrency
+
+                return cell
+            } else {
+                let cell = tableView.dequeueReusableCell(withIdentifier: "productCancelCell", for: indexPath) as! InfoCell
+
+                cell.titleLabel.text = quote?.products[indexPath.row].title
+                let percentageFee = quote!.percentageFee(product: quote!.products[indexPath.row])
+                cell.valueLabel.text = "Cancellation Fee \(percentageFee)%"
+                cell.secondValueLabel?.text = quote?.products[indexPath.row].cancellationCharge.localizedDescriptionInBaseCurrency
+
+                return cell
+            }
+        } else {
+            let cell = tableView.dequeueReusableCell(withIdentifier: "cancelReasonCell", for: indexPath) as! InfoCell
+            cell.titleLabel.text = quote?.cancellationReasons[indexPath.row].value
+            cell.accessoryType = indexPath == selectedReasonIndexPath ? .checkmark : .none
+
+            return cell
+        }
+    }
+
+    // MARK: UITableViewDelegate
 
-        cell.titleLabel.text = quote?.products[indexPath.row].title
-        let percentageFee = quote!.percentageFee(product: quote!.products[indexPath.row])
-        cell.valueLabel.text = "Cancellation Fee \(percentageFee)%"
-        cell.secondValueLabel?.text = quote?.products[indexPath.row].cancellationCharge.localizedDescriptionInBaseCurrency
+    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        guard indexPath.section == 1 else {
+            return
+        }
+
+        let previousIndexPath = selectedReasonIndexPath
+        selectedReasonIndexPath = indexPath
+        selectedReason = quote?.cancellationReasons[indexPath.row]
 
-        return cell
+        let required = (selectedReason?.explanationRequired ?? false) ? "(Required)" : "(Optional)"
+        explanationInputTextField.placeholder = "Please provide an explanation. \(required)"
+
+        tableView.reloadRows(at: [previousIndexPath, selectedReasonIndexPath].compactMap({$0}), with: .automatic)
     }
 
     @IBAction func didConfirm(_ sender: Any) {
-        if let quote = quote {
+        guard let quote = quote else {
+            Log("Cancellation quotet not found.", data: nil, level: .error)
+            return
+        }
+
+        guard let reason = selectedReason else {
+            let alert = UIAlertController(title: "Cancellation reason not selected", message: nil, preferredStyle: .alert)
+            let okAction = UIAlertAction(title: "Ok", style: .cancel, handler: nil)
+            alert.addAction(okAction)
+            present(alert, animated: true, completion: nil)
+
+            return
+        }
+
+        do {
+            let cancellationRequest = try CancellationRequest(quote: quote, reason: reason, explanation: explanation)
```

Comment:
This kind of error handling does not belong here. It belongs to the delegates. When initializing you should be able to do so without having to do error handling. If the initialization requires validation then you do an optional initializer. Will talk more in depth about this on the core PR.

Reply:
It needs some discussion.